crates/file/src/file_reader.rs

```diff
@@ -0,0 +1,37 @@
+use std::rc::Rc;
+
+use futures::Future;
+use wasm_bindgen::JsCast;
+
+use crate::blob::Blob;
+
+pub struct FileReader {
+    inner: web_sys::FileReader,
+}
+
+impl FileReader {
+    pub fn new() -> FileReader {
+        FileReader {
+            inner: web_sys::FileReader::new().unwrap(),
+        }
+    }
+
+    pub fn read_as_string(
+        self,
+        blob: &impl Blob,
+    ) -> impl futures::Future<Item = String, Error = ()> {
+        let (tx, rx) = futures::sync::oneshot::channel();
+        let reader = Rc::new(self.inner);
+        let cloned_reader = reader.clone();
```

Comment:
Cloning the `inner` should be ~about as fast as `Rc` cloning, but won't force us to include all the `Rc` code in the wasm binary.